Add support for Instant as an input and output attribute.

danielgmyers opened the pull request:
* Add support for Instant as an input and output attribute.
* Improve the unit tests for the StepAttributes class.

https://github.com/awslabs/flux-swf-client/issues/19



By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

danielgmyers commented:
The change as it is right now stores Instants as a timestamp in millis, losing precision.

The other approach I've been considering is to allow the attribute provider to provide either type but Flux would store it as provided, and convert it to the step's desired type as needed.  So an Instant requested as a Date would lose the nanos (same behavior as the code change as it is right now), and a Date requested as an Instant would just have 0s for the nanos (also same behavior as the code change as it is right now), but an Instant requested as an Instant would keep the precision.

Also I just realized the WorkflowGraphBuilder's graph validation code for attribute availability needs to treat Date and Instant as interchangeable, so that's missing from this change and will need to be added anyway.

akhomchenko commented:
Have you considered `jackson-datatype-jsr310`? You can use it for more types then without a need for manual conversion:

```
public class JacksonInstantSample {
    public static void main(String[] args) throws IOException {
        ObjectMapper mapper = new ObjectMapper()
                .registerModule(new JavaTimeModule());
                /* alternative to enabling at read/write: enable on top level
                .disable(SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS)
                .disable(DeserializationFeature.READ_DATE_TIMESTAMPS_AS_NANOSECONDS);
                 */

        Instant now = Instant.now().minus(10, ChronoUnit.MINUTES);
        Date date = Date.from(now);

        ObjectWriter writingWithoutNanos = mapper.writer()
                .without(SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS);
        String nowRepr = writingWithoutNanos.writeValueAsString(now);
        String dateRepr = writingWithoutNanos.writeValueAsString(date);

        System.out.println("written values are equal: " + nowRepr.equals(dateRepr) + " and are: " + nowRepr);

        ObjectReader readingWithoutNanos = mapper.reader()
                .without(DeserializationFeature.READ_DATE_TIMESTAMPS_AS_NANOSECONDS);
        Date dateReadBack = readingWithoutNanos.readValue(nowRepr, Date.class);
        Instant instantReadBack = readingWithoutNanos.readValue(dateRepr, Instant.class);

        System.out.println("read values are equal: " + instantReadBack.equals(dateReadBack.toInstant()));

        ZonedDateTime zonedWithTZDefaultForMapper = readingWithoutNanos.readValue(dateRepr, ZonedDateTime.class);
        System.out.println(zonedWithTZDefaultForMapper);

        // if context knows TZ

        ZonedDateTime zonedWithKnownTZ = readingWithoutNanos
                .with(TimeZone.getTimeZone("America/Los_Angeles"))
                .without(DeserializationFeature.ADJUST_DATES_TO_CONTEXT_TIME_ZONE)
                .readValue(dateRepr, ZonedDateTime.class);
        System.out.println(zonedWithKnownTZ);
    }
}
```

danielgmyers commented:
@akhomchenko Interesting; if I stay with the solution that loses nanoseconds precision, I'll try that out.

danielgmyers commented:
I discussed this with @joeymarylander ; we think nanosecond-level precision is not meaningfully useful for timestamps once you're leaving the machine that generated it, especially considering SWF's timers have only 1-second granularity.

Additionally, since our WorkflowGraphBuilder validates the specific input types requested by the workflow steps, and since Date is being superseded by Instant in general, we think it makes the most sense to just add specific support for Instant at the framework level. Users can convert from Instant to another type inside the step if necessary.

danielgmyers commented:
I've updated the change as follows:

* Updated `WorkflowGraphBuilder` to treat Instant and Date as interchangeable.
* Removed internal uses of `OffsetDateTime`. This was a historical artifact from when we migrated off of Joda DateTime, and then after we upgraded to AWS SDK v2 all of these OffsetDateTimes ended up just wrapping the Instants provided directly by the SDK. We were then comparing them as Instants or as millisecond timestamps in all cases.

joeymarylander commented on flux/src/test/java/software/amazon/aws/clients/swf/flux/wf/graph/WorkflowGraphBuilderTest.java:
```diff
@@ -1441,6 +1442,26 @@ public void build_DoesValidateAttributeInputsIfInitialInputAttributesSpecified_A
+
+    @Test
+    public void build_DoesValidateAttributeInputsIfInitialInputAttributesSpecified_AllowsStartTimeAttributesAsInstant() {
+        WorkflowStep start = new TestStepExpectsStartTimeAttributesAsDate();
```
you meant for this to be the instant class, not the Date class.

danielgmyers replied:
good catch, thanks.
